File: ninloader/ninloader/core/validator.py
"""JWT decoding and token expiry checking — zero external dependencies."""

from __future__ import annotations

import base64
import json
import time
from datetime import datetime, timezone
from typing import Any, Dict, Optional, Tuple

from ..types import CollectedToken, ValidationResult


def decode_jwt_payload(token: str) -> Optional[Dict[str, Any]]:
    """Decode a JWT payload without signature verification.

    This is intentional — we're inspecting tokens we already possess,
    not verifying tokens from untrusted sources.
    """
    try:
        parts = token.split(".")
        if len(parts) != 3:
            return None
        # Add padding
        payload_b64 = parts[1]
        padding = 4 - len(payload_b64) % 4
        if padding != 4:
            payload_b64 += "=" * padding
        payload_bytes = base64.urlsafe_b64decode(payload_b64)
        return json.loads(payload_bytes)
    except Exception:
        return None


def check_expiry(token: CollectedToken) -> Tuple[bool, Optional[str]]:
    """Check if a token is expired.

    Returns (expired: bool, expires_at: str|None).
    Checks token.expires_at first, then attempts JWT exp claim.
    """
    now = time.time()

    # Check explicit expires_at
    if token.expires_at:
        try:
            exp_dt = datetime.fromisoformat(token.expires_at.replace("Z", "+00:00"))
            expired = exp_dt.timestamp() < now
            return expired, token.expires_at
        except (ValueError, AttributeError):
            pass

    # Try JWT exp claim from access_token
    if token.access_token:
        payload = decode_jwt_payload(token.access_token.value)
        if payload and "exp" in payload:
            exp_ts = payload["exp"]
            exp_dt = datetime.fromtimestamp(exp_ts, tz=timezone.utc)
            expired = exp_ts < now
            return expired, exp_dt.isoformat()

    return False, None


def validate_token(token: CollectedToken) -> ValidationResult:
    """Validate a collected token — check structure and expiry."""
    # Must have at least one credential
    has_access = token.access_token and token.access_token.value
    has_refresh = token.refresh_token and token.refresh_token.value
    has_secret = token.client_secret and token.client_secret.value

    if not has_access and not has_refresh and not has_secret:
        return ValidationResult(
            valid=False,
            service=token.service,
            source=token.source,
            account_hint=token.username,
            error="No token material found",
        )

    expired, expires_at = check_expiry(token)

    # If only a refresh token, it's valid even if access is expired
    if expired and has_refresh:
        return ValidationResult(
            valid=True,
            service=token.service,
            source=token.source,
            account_hint=token.username,
            expires_at=expires_at,
            expired=True,
            error="Access token expired but refresh token available",
        )

    if expired:
        return ValidationResult(
            valid=False,
            service=token.service,
            source=token.source,
            account_hint=token.username,
            expires_at=expires_at,
            expired=True,
            error="Token expired",
        )

    return ValidationResult(
        valid=True,
        service=token.service,
        source=token.source,
        account_hint=token.username,
        expires_at=expires_at,
    )

File: ninloader/go/internal/validator/validator.go
package validator

import (
	"encoding/base64"
	"encoding/json"
	"strings"
	"time"

	"github.com/ninken/ninloader-go/internal/types"
)

// DecodeJWTPayload decodes a JWT without signature verification.
// Returns nil if not a valid JWT.
func DecodeJWTPayload(token string) map[string]any {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil
	}
	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil
	}
	var result map[string]any
	if err := json.Unmarshal(payload, &result); err != nil {
		return nil
	}
	return result
}

// CheckExpiry checks if a token is expired.
// Returns (expired, expiresAt string).
func CheckExpiry(token *types.CollectedToken) (bool, string) {
	now := time.Now().UTC()

	// Check explicit expires_at
	if token.ExpiresAt != "" {
		expDt, err := time.Parse(time.RFC3339, token.ExpiresAt)
		if err != nil {
			// Try ISO format without timezone
			expDt, err = time.Parse("2006-01-02T15:04:05", token.ExpiresAt)
		}
		if err == nil {
			return expDt.Before(now), token.ExpiresAt
		}
	}

	// Try JWT exp claim from access_token
	if token.AccessToken != nil && !token.AccessToken.IsEmpty() {
		payload := DecodeJWTPayload(token.AccessToken.Value())
		if payload != nil {
			if exp, ok := payload["exp"].(float64); ok {
				expTime := time.Unix(int64(exp), 0).UTC()
				return expTime.Before(now), expTime.Format(time.RFC3339)
			}
		}
	}

	return false, ""
}

// ValidateToken checks structure and expiry.
func ValidateToken(token *types.CollectedToken) *types.ValidationResult {
	hasAccess := token.AccessToken != nil && !token.AccessToken.IsEmpty()
	hasRefresh := token.RefreshToken != nil && !token.RefreshToken.IsEmpty()
	hasSecret := token.ClientSecret != nil && !token.ClientSecret.IsEmpty()

	if !hasAccess && !hasRefresh && !hasSecret {
		return &types.ValidationResult{
			Valid:       false,
			Service:     token.Service,
			Source:      token.Source,
			AccountHint: token.Username,
			Error:       "No token material found",
		}
	}

	expired, expiresAt := CheckExpiry(token)

	if expired && hasRefresh {
		return &types.ValidationResult{
			Valid:       true,
			Service:     token.Service,
			Source:      token.Source,
			AccountHint: token.Username,
			ExpiresAt:   expiresAt,
			Expired:     true,
			Error:       "Access token expired but refresh token available",
		}
	}

	if expired {
		return &types.ValidationResult{
			Valid:       false,
			Service:     token.Service,
			Source:      token.Source,
			AccountHint: token.Username,
			ExpiresAt:   expiresAt,
			Expired:     true,
			Error:       "Token expired",
		}
	}

	return &types.ValidationResult{
		Valid:       true,
		Service:     token.Service,
		Source:      token.Source,
		AccountHint: token.Username,
		ExpiresAt:   expiresAt,
	}
}
